ibtk/src/utilities/PartitioningBox.cpp

```diff
@@ -0,0 +1,124 @@
+#include <ibtk/ibtk_utilities.h>
+#include <ibtk/PartitioningBox.h>
+
+#include <CartesianPatchGeometry.h>
+
+#include <algorithm>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+namespace IBTK
+{
+PartitioningBox::PartitioningBox()
+{
+    for (unsigned int dim_n = 0; dim_n < NDIM; ++dim_n)
+    {
+        d_bounding_points.first[dim_n] = std::numeric_limits<double>::infinity();
+        d_bounding_points.second[dim_n] = std::numeric_limits<double>::infinity();
+    }
+} // PartitioningBox
+
+PartitioningBox::PartitioningBox(const Point &bottom_point,
+                                 const Point &top_point)
+    : d_bounding_points(bottom_point, top_point)
+{
+    for (unsigned int dim_n = 0; dim_n < NDIM; ++dim_n)
+    {
+        // Do not permit negative volume boxes
+        TBOX_ASSERT(bottom()[dim_n] <= top()[dim_n]);
+    }
+} // PartitioningBox
+
+PartitioningBox::PartitioningBox(const SAMRAI::geom::CartesianPatchGeometry<NDIM> &patch)
+{
+    std::copy(patch.getXLower(), patch.getXLower() + NDIM,
+              d_bounding_points.first.data());
+    std::copy(patch.getXUpper(), patch.getXUpper() + NDIM,
+              d_bounding_points.second.data());
+    for (unsigned int dim_n = 0; dim_n < NDIM; ++dim_n)
+    {
+        // Do not permit negative volume boxes
+        TBOX_ASSERT(bottom()[dim_n] <= top()[dim_n]);
+    }
+} // PartitioningBox
+
+double PartitioningBox::volume() const
+{
+    // The bounds may be infinite: check for point equality first to avoid
+    // evaluating oo - oo
+    for (unsigned int dim_n = 0; dim_n < NDIM; ++dim_n)
+    {
+        if (d_bounding_points.first[dim_n] == d_bounding_points.second[dim_n])
+            return 0.0;
+    }
+    double vol = 1.0;
+    for (unsigned int dim_n = 0; dim_n < NDIM; ++dim_n)
+    {
+        vol *= d_bounding_points.second[dim_n] - d_bounding_points.first[dim_n];
+    }
+    return vol;
+} // volume
+
+namespace
+{
+    std::vector<IBTK::PartitioningBox>
+    get_partitioning_boxes(const SAMRAI::hier::PatchHierarchy<NDIM> &patch_hierarchy)
+    {
+        std::vector<IBTK::PartitioningBox> boxes;
+        const int finest_level = patch_hierarchy.getFinestLevelNumber();
+        SAMRAI::tbox::Pointer<SAMRAI::hier::PatchLevel<NDIM> > level
+            = patch_hierarchy.getPatchLevel(finest_level);
+        for (SAMRAI::hier::PatchLevel<NDIM>::Iterator p(level); p; p++)
+        {
+            const SAMRAI::hier::Patch<NDIM>& patch = *level->getPatch(p());
+            SAMRAI::tbox::Pointer<SAMRAI::geom::CartesianPatchGeometry<NDIM> > patch_geometry
+                = patch.getPatchGeometry();
+            boxes.emplace_back(*patch_geometry);
+        }
+        return boxes;
+    }
+}
+
+PartitioningBoxes::PartitioningBoxes(const SAMRAI::hier::PatchHierarchy<NDIM> &hierarchy)
+{
+    const auto boxes = get_partitioning_boxes(hierarchy);
```

Comment:
Not sure that this use of `auto` follows the unofficial rules-of-auto.

Reply:
This is easy to redo. I'll fix it.